Sparse softmax support (CPU)

pearu opened the pull request:
This PR implements softmax support for sparse tensors.

The sparse softmax is related to dense softmax when the values of unspecified sparse tensor entries are taken to be `-inf` that will have the effect of "zero entries ignored". This relation is used for testing the correctness of results here.

Resolves #23651 for CPU.

- [x] sparse softmax
  - [x] CPU C++ implementation
  - [x] unittests
  - [x] update softmax documentation
  - [x] autograd support
- [x] sparse log_softmax
  - [x] CPU C++ implementation
  - [x] unittests
  - [x] update log_softmax documentation
  - [x] autograd support

pearu committed:
Prepare for softmax_sparse_cpu

pearu committed:
Implement sparse softmax in Python to test the correctness of the algorithm.

pearu committed:
Implemented C++ sparse softmax

pearu committed:
Implement sparse log_softmax for CPU

pearu committed:
Save some division cycles

pearu committed:
Implement autograd support for sparse softmax

pearu committed:
Implement autograd support for sparse log_softmax

pearu committed:
Skip test_softmax on non-cpu devices

pearu committed:
Add a sparse tensor fill value note to softmax docs.

ezyang commented:
> The sparse softmax is related to dense softmax when the values of unspecified sparse tensor entries are taken to be -inf that will have the effect of "zero entries ignored"

We had a bunch of discussions about whether or not it is appropriate to treat unspecified sparse tensor entries as things other than zero; specifically, whether or not we should be allowed to overload the old function name in this situation (some discussion about this at https://github.com/pytorch/pytorch/issues/1369 ). Did you folks at Quansight come to a decision about what should be done here?

pearu commented:
@ezyang we have not discussed the overloading vs name-prefixing options yet but we have discussed the idea of introducing the so-called fill value to sparse tensors that would set the default value for unspecified sparse tensor entries.

The fill value feature would be useful also for the sparse softmax case that would allow users to specify the user-defined behavior:
- Using `fill_value=0` would lead to nice invariant `softmax(sparse, dim).to_dense() == softmax(sparse.to_dense(), dim)` while the result of `softmax(sparse, dim)` would be still sparse tensor (with a non-zero `fill_value`).
- Using `fill_value=-inf` [the current PR] gives a mathematical basis for `softmax(sparse, dim)` to be a sparse tensor (as requested in the issue #23651).

In addition, the fill value feature has the potential to eliminate the overloading vs name-prefixing question, and in general, it would lower the bridge between sparse and dense tensor in a consistent way.

The fill value has been implemented in https://sparse.pydata.org/ and from https://github.com/pytorch/pytorch/issues/9674#issuecomment-406822741 I see you have already discussed this option as well.

For this PR, we have two choices:
1. sparse softmax is exposed as  `softmax` (and `log_softmax`) [the current behavior] via overloading
2. sparse softmax is exposed as `spsoftmax` (and `splog_softmax`)

In the long term, with the perspective of having the fill value feature available, option 1. makes sense to me.
On the other hand, option 2. seems to be the current practice in pytorch and I would be ok changing the PR accordingly.

So, I am not certain atm which approach would be preferable to users. Thoughts?

ezyang commented:
If y'all are prepared to take on the maintenance burden of a fill value, I think I would also agree that this would be best, and then we don't have to introduce sparse versions of all of our functions. Without fill, (1) will technically require you to do BC-breaking changes in the future, as you will turn an operation that previously implicitly converted zero-filled tensors to neginf-filled tensors, into an error. It will save you work in the future if you do (2) first.

pearu commented:
OK, it makes sense. Here's a plan:
- [ ] Expose sparse softmax as `torch.sparse.softmax` that is defined for sparse tensors with assumed fill value equal to negative infinity that is equivalent to ignoring unspecified sparse tensor entries.
- [ ] Expose sparse log softmax as `torch.sparse.log_softmax` with fill value as above.
- Defining `torch.softmax` on sparse tensors will be reserved for sparse tensors that support fill value feature. The result of `torch.softmax` will be different from the result of `torch.sparse.softmax` because of the default fill value for sparse tensors will be zero. The results will be equal iff the fill value of a sparse tensor is specified as negative infinity.

ezyang commented:
SGTM.

pearu committed:
Expose sparse softmax via torch.sparse.softmax. Same for log_softmax. Implement support for dtype argument.

pearu committed:
Fix softmax and log_softmax signatures.

pearu committed:
Clean up

pearu commented:
@apaszke @ezyang this PR is ready for review. The PR implements sparse softmax support for CPU, the CUDA support is planned as a followup PR.

ezyang commented:
@pearu can you get someone from Quansight to review first?

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,466 @@
+  iscalar_t mx_p;
+  std::vector<iscalar_t> pool = get_dense_offsets<iscalar_t>(mx_p, offsets);
+
+  /* Compute mx */
+  std::vector<int64_t> mx_sizes(sizes.begin() + sparse_dim - 1, sizes.end());
+  mx_sizes[0] = mx_p;
+  at::Tensor mx = at::empty(mx_sizes, values.options());
+  scalar_t* mx_data_base = mx.data_ptr<scalar_t>();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
```
I am not sure, but are you sure that `max` does not work with sparse tensors?

pearu replied:
You can apply max on sparse tensor values but max is not supported on SparceCPU backend, for instance.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,466 @@
+  }
+
+  /* normalize with the sum of exponents */
+  for (int64_t i=0; i < nnz; i++) {
+    auto p = pool[i];
+    scalar_t* values_data = values_data_base + i * nvalues;
+    scalar_t* out_values_data = out_values_data_base + i * nvalues;
+    scalar_t* exp_sums_data = exp_sums_data_base + p * nvalues;
+    scalar_t* mx_data = mx_data_base + p * nvalues;
+    for (int64_t j=0; j < nvalues; j++) {
+      if (LogSoftMax) {
+        out_values_data[j] = values_data[j] - mx_data[j];
+      } else {
+        out_values_data[j] *= exp_sums_data[j];
+      }
+    }
+  }
```
I would use TensorIterator here.

nikitaved replied:
You could also use accessors to eliminate raw pointer exposures.

pearu replied:
I have found no evidence that using TensorIterator would perform better here.
Even when assuming `nvalues << nnz`, the need to use the `pool` mapping makes the current loops most effective for the considered test cases (see `test/test_sparse.py::test_softmax`).

nikitaved replied:
Could you please show how you used TensorIterator here?

nikitaved replied:
And I mean the loop over `nnz`, not `nvalues`.

pearu replied:
@nikitaved , `cpu_sparse_coo_softmax` has five for-loops, only the last loop is independent in writes for different `i in range(nnz)`, and the first and the 4rd are independent in writes for different `i in range(npools)`. In most cases, `npools << nnz`.

pearu replied:
Here are cumulative timings for the five loops when running `test_softmax`:
```
loop 1 took     366590 ns
loop 2 took     329048 ns
loop 3 took    2075724 ns
loop 4 took     774612 ns
loop 5 took     349260 ns
```
As you can see, the 3rd loop dominates, most likely due to the `exp` calls.
The 4th loop has also `log` call but this loop over `range(npools)` is short compared to the loop over `range(nnz)`.

pearu replied:
It is possible to refactor the loop over `range(nnz)` to independent sub-loops.
Consider the 3rd loop (for simplicity, using scalar dense part):
```
for (int64_t i=0; i < nnz; i++) {
  auto p = pool[i];
  auto v = exp(values[i] - mx[p]);
  out_values[i] = v;
  exp_sums[p] += v;
}
```
that can be replaced with
```
for (int64_t p=0; p < npools; p++) {
  auto mx_p = mx[p];
  scalar_t acc = 0;
  for (int64_t i=0; i < nnz; i++) {
    if (pool[i] == p) {
      auto v = exp(values[i] - mx_p);
      out_values[i] = v;
      acc += v;
    }
  }
  exp_sums[p] = acc;
}
```
so that the loops over `p` are now independent and could be run in parallel.

Can you suggest how to set up a TensorIterator for the above?

pearu replied:
Notice that with the above loop-refactoring all five loops could be joined under one loop over `p in range(npools)` which makes the approach very interesting.

pearu replied:
All loops are now restructured to use independent pools of indices (=> increase of performance about 25%) and using parallel_for (=> additional increase of performance about 7-8%).

nikitaved commented:
All the loops could be replaced with the TensorIterator loops, I think...

pearu committed:
Minor clean up.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+  return offsets;
+}
+
+template <typename iscalar_t>
+std::vector<iscalar_t> get_dense_offsets(iscalar_t &mx, const std::vector<iscalar_t>& offsets) {
+  /*
+    Return the dense set of offsets:
+
+      If
+        pool = get_dense_offsets(mx, offsets)
+      then
+        pool.size() == offsets.size()
+        pool[i] == pool[j] iff offsets[i] == offsets[j]
+        set(pool) == set(range(mx))
+  */
```
nit: Not clear to me what this function is doing, there is also a `get_offsets` which also returns offsets in a dense tensor...

pearu replied:
Updated the docs with more details and an example.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+namespace {
+
+template <typename iscalar_t>
+iscalar_t get_nvalues(const std::vector<iscalar_t>& sizes, const int64_t sparse_dim) {
+  /* Return the number of elements of sparse values entries.
+   */
+  iscalar_t nvalues = 1;
+  for (auto it = sizes.begin() + sparse_dim; it != sizes.end(); ++it) {
+    nvalues *= *it;
+  }
+  return nvalues;
+}
```
nit: Also the description is not clear to me. Looks like you compute the number of elements from the specified dim to the end?

pearu replied:
Yes, in sparse tensors the dimensions are "sorted": the first part of `sizes` corresponds to sparse dimensions and the rest to dense dimensions. The `get_nvalues` computes the number of entries of the dense part of the sparse tensor.

pearu replied:
Updated the description.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+  return pool;
+}
+
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
```
nit: I think that index is always of type `int64_t`.

pearu replied:
renamed `iscalar_t` to `index_t`.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+  auto sizes = input.sizes().vec();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Compute pool indices */
+  std::vector<iscalar_t> offsets = get_offsets<iscalar_t>(indices, sizes, dim);
+  iscalar_t mx_p;
+  std::vector<iscalar_t> pool = get_dense_offsets<iscalar_t>(mx_p, offsets);
```
nit: Could you please insert more explanation about this `pool` mapping?

pearu replied:
Yes, the sparse/SoftMax.cpp now contains the derivation of the sparse softmax algorithm with an example.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+}
+
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
+void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, const Tensor& output, const int64_t dim) {
+  /*
+    gI_i = sum_j d<output_i>/d<input_j> * grad_j = sum_j output_i * (1[i==j] - output_j) * grad_j
+         = output_i * (grad_i - sum_j output_j * grad_j)
+
+    i, j in range(shape[dim])
+    x_i = x[..., i_dim, ...]
+
+    assuming output.sparse_dim() == grad.sparse_dim(), TODO: impl check
+  */
```
Is `output` the result of softmax?
Suppose this is the case, then
```
grad_input_i = 1[i==j] output_i * grad_i - output_i * output_j * grad_j
```

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+template <typename scalar_t, typename iscalar_t, bool LogSoftMax>
+void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, const Tensor& output, const int64_t dim) {
+  /*
+    gI_i = sum_j d<output_i>/d<input_j> * grad_j = sum_j output_i * (1[i==j] - output_j) * grad_j
```
`gI_i` = sum_j d<output_j>/d<input_i>`

nikitaved replied:
And the formula is wrt to LogSoftMax=false in here.

nikitaved replied:
And the formula below is for LogSoftMax=false.

pearu replied:
Right, fixed. Thanks!

nikitaved commented:
Forward looks good to me module `pool` mapping which I am not sure I understand. Could you please put some description about it's purpose?

I still have to look at the backward...

@ezyang , there is no TensorIterator nor parallel_for loop. Is that still all right for?

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, const Tensor& output, const int64_t dim) {
+  /*
+    gI_i = sum_j d<output_i>/d<input_j> * grad_j = sum_j output_i * (1[i==j] - output_j) * grad_j
+         = output_i * (grad_i - sum_j output_j * grad_j)
```
This whole thing gets premultiplied with `1/output_j` when LogSoftMax=true.

nikitaved commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+void cpu_sparse_coo_softmax_backward(Tensor& grad_input, const Tensor& grad, const Tensor& output, const int64_t dim) {
+  /*
+    gI_i = sum_j d<output_i>/d<input_j> * grad_j = sum_j output_i * (1[i==j] - output_j) * grad_j
+         = output_i * (grad_i - sum_j output_j * grad_j)
+
+    i, j in range(shape[dim])
+    x_i = x[..., i_dim, ...]
```
Valid when LogSoftMax = false.
When LogSoftMax = true it should be:
`grad_input_i = (1-exp(output_i)) * grad_i - sum_{j!=i} exp(output_i) * exp(output_j) * grad_j`

pearu replied:
Added to the comment. Thanks!

nikitaved commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3255,6 +3255,40 @@
+  dispatch:
+    SparseCPU: softmax_sparse_cpu
+
+- func: _sparse_softmax_backward_data(Tensor grad_output, Tensor output, int dim, Tensor self) -> Tensor
+  dispatch:
+    SparseCPU: softmax_backward_sparse_cpu
```
@albanD , could you please tell whether backward for a backward is required here?

albanD replied:
Given that the backaward implementation cannot be automatically differentiated, you will get an error if you try to differentiate through this.
In general, we don't block a feature on not having a double backward.
If you want it here though, you can define a formula for it as `softmax_backward_backward_sparse_cpu` in `tools/autograd/template/Functions.cpp` that only uses high level pytorch ops so that it will be infinitely differentiable.

pearu replied:
With the sparse tensor limited support for high level pytorch operations, I suggest postponing the double backward for sparse softmax.

nikitaved reviewed (approved):
LGTM. maybe some parallel_for for performance.
Backwards for backwards would be great to have. The dense version does support it.

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+  auto result = [&]() {
+    NoNamesGuard guard;
+    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+        return at::_sparse_log_softmax(input_, dim_, true);
```
Hmm, what's going on here?

pearu replied:
This is a shameless copy from aten/src/Aten/native/SoftMax.cpp. For explanation, when a user requests output dtype to be different from input dtype, by default, the input is converted to output dtype, and the softmax is applied to the converted input. However, for Half input & Float output case, the softmax cuda implementation keeps the input as it is (Half dtype) and the results are accumulated to (Float dtype) output. As I understand it, this approach allows saving CUDA device memory.

hameerabbasi commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+    if (input_.is_cuda() && input_.scalar_type() == ScalarType::Half && dtype == ScalarType::Float){
+        return at::_sparse_log_softmax(input_, dim_, true);
+    } else {
+        Tensor converted = dtype.has_value() ? input_.toType(dtype.value()) : input_;
```
Is it at all possible to compute with the right `dtype` upfront?

pearu replied:
The current approach may not be optimal indeed for the general case.
Only the half-input to float-output or output-dtype-is-same-as-of-input cases are handled optimally. In this PR, we'll use the same approach as used in the dense softmax implementation for consistency.

hameerabbasi commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,467 @@
+  TORCH_CHECK(
+              grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
```
I guess the `dtype` dispatching would take place here.

pearu replied:
yes

hameerabbasi reviewed (commented):
LGTM, from an algorithm point of view. I have one small performance-related comment.

pearu committed:
Derivation of the sparse softmax algorithm with an example

pearu committed:
Use accessors instead of raw pointers. Use 2-D views of the (N+1)-D values. Rename iscalar_t to index_t. Update docs. Clean up.

pearu committed:
Refactor to use loops over independent pools. Increase of performance is about 25%.

pearu committed:
Using parallel_for, increase of performance about 8 %.

pearu commented:
I have addressed the review comments as follows:
- Documented the derivation of the algorithm used in sparse softmax, this includes the explanation for the notion of pools of indices.
- Restructured softmax loops to use independent pools of indices so that paraller_for can be used. The restructuring alone increased the sparse softmax performance by about 25%, parallel_for added 7% of increase (based on softmax unit tests).
- Using accessors instead of raw pointers.
- Postpone double backward implementation because of the limited support of high-level operations for sparse tensors.

Please review.

ezyang commented on torch/sparse/__init__.py:
```diff
@@ -137,3 +139,55 @@ def sum(input, dim=None, dtype=None):
+
+
+def softmax(input, dim, dtype=None):
+    # type: (Tensor, int, Optional[int]) -> Tensor
```
we python 3 only, can do these annotations inline!

pearu replied:
Fixed.

ezyang commented on torch/sparse/__init__.py:
```diff
@@ -137,3 +139,55 @@ def sum(input, dim=None, dtype=None):
+    if dtype is None:
+        return torch._sparse_softmax(input, dim)
+    else:
+        return torch._sparse_softmax(input, dim, dtype=dtype)
```
Weird that we can't just call `_sparse_softmax` with `dtype=None` here

pearu replied:
It turns out that we can. Fixed.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
         t = torch.sparse_coo_tensor(torch.tensor(([0, 0], [2, 0])), torch.tensor([1, 4]))
         self.assertRaises(TypeError, lambda: t.numpy())
 
+    def test_softmax(self):
```
You could use `@cpu_only` decorator here

pearu replied:
yes, fixed.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+            """
+            Return dense tensor from a sparse tensor using given fill value.
+            """
+            if isinstance(fill_value, str):
```
lolwut

pearu replied:
yeah, that was used for `fill_value=float("-inf")`, now removed.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+                return sparse.to_dense()
+            sparse = sparse.coalesce()
+            dense = torch.empty(sparse.shape, dtype=sparse.dtype, device=sparse.device)
+            dense[:] = fill_value
```
You can do this and the line above in one go using `torch.full`, which takes a fill value.

pearu replied:
Fixed. Thanks for the hint!

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+            dense = torch.empty(sparse.shape, dtype=sparse.dtype, device=sparse.device)
+            dense[:] = fill_value
+            for idx, value in zip(sparse._indices().t(), sparse._values()):
+                dense[tuple(idx)] = value
```
This is going to be soooo sloooow. Probably OK for testing, but I bet you can probably do this faster by scattering zeros to the locations of `dense` referenced by `indices`, and then doing a dense-sparse addition. NO ACTION NECESSARY ON THIS COMMENT.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+            """
+            dtype = sparse.dtype
+            device = sparse.device
+            sparse = sparse.coalesce()
```
I don't think this coalesce should be necessary

pearu replied:
Right, `to_dense` calls `coalesce`. Fixed.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+            sparse = sparse.coalesce()
+            dense = to_dense(sparse, fill_value=-float('inf'))
+            r = F.softmax(dense, dim)
+            # softmax on empty lines results nan, replace with zeros to match the definition
```
Q: if we got sparse tensors with fills, what would the fill of the output of sparse softmax be?

pearu replied:
Let `X` be an `N`-D sparse tensor with fill value `x0`, `Y = softmax(X, d)`. Consider a single line along the dimension d at `i0, i1, ...`:
```
x=X[i0, i1, ..., range(X.shape[d]), ...]
```
that contains `n` specified entries with indices in `P`, and `m` unspecified entries (`n + m == X.shape[d]`). Then the fill value for this particular line is
```
y0 = exp(x0) / (m * exp(x0) + sum(exp(x[i]), i in P))
```
Clearly, the value of `y0` will vary for different `i0, i1, ...`.

In general, there will be `prod(X.shape) // X.shape[d]` different "fill values" for `Y` that can be collected in a `N`-D tensor, say `Z`, with complemented indices to `X` (and `Y`). I think, `Z` will be isomorphic to a `(N-1)`-D dense tensor. By introducing `Z`, the single fill value of `Y` would be `0`.

So, the answer will be `0` but with the introduction of a complementary tensor `Z` that entries approach to `0` as `x0 -> -inf`.

ezyang commented on test/test_sparse.py:
```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+        test_op(3, 100, [3, 4, 2])
+        test_op(3, 100, [3, 4, 2, 3])
+        test_op(3, 100, [3, 4, 2, 3, 5, 2])
+        test_op(4, 100, [3, 4, 2, 3, 5, 2])
```
How long do these tests take to run?

pearu replied:
A single run of `test_softmax` takes about 0.5 seconds.

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+  }
+
+  auto nnz = values.size(0);
+  auto sizes = input.sizes().vec();
```
You don't seem to write entries into the sizes vector here... do you really need to materialize a `vec()` here? If you drop the `vec()` you'll get an `IntArrayRef` which is a view type into the sizes. You'd then have to change `get_nvalues` to take `IntArrayRef` rather than `const std::vector<int>&` but that should be fine.

pearu replied:
Yes, good point. Fixed.

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+    }
+    while (pool_index >= pools.size()) {
+      std::vector<index_t> indices;
+      pools.push_back(indices);
```
I think `pools.push_back()` would be equivalent here.

pearu replied:
It is `pools.emplace_back()`. Fixed.

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+}
+
+template <typename index_t>
+std::vector<std::vector<index_t>> get_pools(const Tensor& indices, const std::vector<index_t>& sizes, const int64_t dim) {
```
I guess this returns a vector of vectors because it's ragged?

pearu replied:
yes, exactly.

ezyang commented:
Any performance numbers?

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+    The implementation below has more optimizations such as that
+    collect pool indices for enabling concurrency, minimize the calls
+    to exp functions as well as reuse of softmax implementation for
+    log_softmax.
```
Thank you, this is a very nice comment. I wish every kernel had comments like this :)

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+} // namespace
+
+Tensor softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
+  AT_ASSERT(input_.is_sparse());
```
The proper modern macro is `TORCH_INTERNAL_ASSERT` here (for history for the rename, see https://github.com/pytorch/pytorch/issues/20287

pearu replied:
Fixed.

ezyang commented on aten/src/ATen/native/sparse/SoftMax.cpp:
```diff
@@ -0,0 +1,647 @@
+
+Tensor softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
+  AT_ASSERT(input_.is_sparse());
+  AT_ASSERTM(!half_to_float, "softmax with half to float conversion is not supported on CPU");
```
Building on the above, this one should be a `TORCH_CHECK`, since a user could reasonably run into this bug (right?)

pearu replied:
Fixed.

ezyang commented:
I pinged @colesbury to take a look at the high level algorithmics. One thing I was wondering a bit about was how much of the infrastructure in this PR might be generalizable into some sort of SparseTensorIterator. It seems to me that we are likely to end up in a regime where we want to do many sparse reductions, and much of the same machinery introduced in this PR is likely to be applicable somewhere else. This doesn't mean we have to go ahead and generalize *right now*, but it means that now might be a good time to make sure (1) that we're not duplicating work in the rest of the sparse kernels and (2) to think about how to generally approach reductions in sparse tensors.

Some thoughts:
1. Does it make sense to have a sort-by-specific-dim variant of coalesce? If I understand correctly, you get good locality in the event that the reduction is done with `dim=0` (because the rows are laid out contiguously) and bad locality in all other cases. If coalescing also took care of transposing the sparse data so that the relevant dimension was all put together, then you could do away with the pooling data structures and just iterate sequentially until you hit the index bump in the dimension.
2. You currently parallelize over pools. This will work ok if the sparse entries are uniformly distributed, but if you have a big imbalance (some pools are really big) you will end up with a bunch of threads that have run out of work early. In general this sort of heterogenous parallelism is difficult to setup, since if you cut the data homogenously (supposing you had done the strategy suggested in (1)) you now have a bunch of fiddly boundary cases to deal with.

ezyang commented:
Re when to land this PR: I'm generally amenable to landing this sooner rather than later, but it would be cool to talk about some of these things so we know where diffs like this are headed in the future.

hameerabbasi commented:
Re: SparseTensorIterator, that's exactly what TACO is essentially. https://github.com/tensor-compiler/taco

pearu committed:
Use inline annotations.

pearu committed:
Eliminate fill_value from str.

pearu committed:
Use IntArrayRef, eliminate usage of index_t, fix ASSERT macro usages.

rgommers commented on torch/sparse/__init__.py:
```diff
@@ -4,10 +4,22 @@
+if TYPE_CHECKING:
+    from torch import dtype as DType
+else:
+    DType = int
```
@ezyang do you happen to know why this is needed? @pearu checked and found that torchscript doesn't support `torch.types._dtype` in the type annotation. That seems like it's going to give problems in other places.

ezyang replied:
Yes, we're going need to clue in TorchScript about this, unfortunately >:(

pearu commented:
Here are some benchmark results for `torch.sparse.softmax` (v1.6.0.a) and `tensorflow.sparse.softmax` (v 2.2.0), both using CPU, values are float64 scalars, timing repeat is 1000:

sizes | nnz | dimension |pytorch sparse | pytorch dense| tensorflow
------| ------|-----|------|--|--
(100, 100) | 1000 | 0 | 127.68 us | 66.16 us|N/A
(100, 100) | 1000 | 1 | 97.27 us |10.36 us| 234.18 us
(100, 100) | 10000 | 0 | 595.16 us |67.35 us| N/A
(100, 100) | 10000 | 1 | 513.72 us |10.22 us| 490.75 us
(100, 100, 100) | 1000 | 0 | 280.73 us |1013.29 us| N/A
(100, 100, 100) | 1000 | 1 | 230.72 us |793.75 us| N/A
(100, 100, 100) | 1000 | 2 | 225.87 us |554.37 us| 324.38 us
(100, 100, 100) | 10000 | 0 |1371.77 us |1135.67 us| N/A
(100, 100, 100) | 10000 | 1 |1228.75 us |691.85 us| N/A
(100, 100, 100) | 10000 | 2 |1171.73 us |533.57 us| 1294.87 us
(100, 100, 100) | 100000 | 0 |7957.96 us |1413.64 us| N/A
(100, 100, 100) | 100000 | 1 |6750.87 us |752.53 us| N/A
(100, 100, 100) | 100000 | 2 |5801.34 us |532.25 us| 6444.39 us

Some notes:
- pytorch.sparse.softmax is generally faster than `tensorflow.sparse.softmax`. That is pretty good considering that pytorch sparse softmax supports more features that tf.
- softmax performance is higher for larger dimension argument, this is likely related to better locality properties of index pools when using row-wise storage.
- pytorch sparse vs dense timings show that there exists a density region (< 1 %) where using sparse storage gives considerably better performance than using dense storage. For relatively dense matrices, usage of sparse storage leads to up to 10x smaller performance.

ezyang commented:
Failure is legit:

```

May 12 18:18:16   test_softmax (__main__.TestSparse) ... /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/SoftMax.cpp:348:35: runtime error: division by zero
May 12 18:18:17     #0 0x7f50670b37cd in .omp_outlined. (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so+0xc2517cd)
May 12 18:18:17     #1 0x7f5053405712 in __kmp_invoke_microtask (/opt/conda/lib/libiomp5.so+0x122712)
May 12 18:18:17     #2 0x7f5053393ffe in __kmp_invoke_task_func (/opt/conda/lib/libiomp5.so+0xb0ffe)
May 12 18:18:17     #3 0x7f5053393059 in __kmp_launch_thread (/opt/conda/lib/libiomp5.so+0xb0059)
May 12 18:18:17     #4 0x7f5053405bd7 in _INTERNALdc942ded::__kmp_launch_worker(void*) (/opt/conda/lib/libiomp5.so+0x122bd7)
May 12 18:18:17     #5 0x7f508a9cd6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)
May 12 18:18:17     #6 0x7f508a70341c in clone /build/glibc-LK5gWL/glibc-2.23/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:109
May 12 18:18:17 
```

pearu committed:
Resolve zero division error issue.

cpuhrsch commented:
Interesting to read the conversations in this PR again in light of [maskedtensor](https://pytorch.org/maskedtensor) cc @george-qi
